// coc.nvim/lib/model/db.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = __importDefault(require("path"));
const fs = __importStar(require("../util/fs"));
class DB {
    constructor(filepath, nvim) {
        this.filepath = filepath;
        this.nvim = nvim;
    }
    async fetch(key) {
        let obj = await this.load();
        if (obj == null)
            return undefined;
        let parts = key.split('.');
        for (let part of parts) {
            if (typeof obj[part] == 'undefined') {
                return undefined;
            }
            obj = obj[part];
        }
        return obj;
    }
    async exists(key) {
        let obj = await this.load();
        if (obj == null)
            return false;
        let parts = key.split('.');
        for (let part of parts) {
            if (typeof obj[part] == 'undefined') {
                return false;
            }
            obj = obj[part];
        }
        return true;
    }
    async delete(key) {
        let obj = await this.load();
        if (obj == null)
            return;
        let origin = obj;
        let parts = key.split('.');
        let len = parts.length;
        for (let i = 0; i < len; i++) {
            if (typeof obj[parts[i]] == 'undefined') {
                break;
            }
            if (i == len - 1) {
                delete obj[parts[i]];
                await fs.writeFile(this.filepath, JSON.stringify(origin, null, 2));
                break;
            }
            obj = obj[parts[i]];
        }
    }
    async push(key, data) {
        let obj = await this.load();
        let origin = obj || {};
        let parts = key.split('.');
        let len = parts.length;
        if (obj == null) {
            let dir = path_1.default.dirname(this.filepath);
            await this.nvim.call('mkdir', [dir, 'p', 0o700]);
            obj = origin;
        }
        for (let i = 0; i < len; i++) {
            let key = parts[i];
            if (i == len - 1) {
                obj[key] = data;
                await fs.writeFile(this.filepath, JSON.stringify(origin, null, 2));
                break;
            }
            if (typeof obj[key] == 'undefined') {
                obj[key] = {};
                obj = obj[key];
            }
            else {
                obj = obj[key];
            }
        }
    }
    async load() {
        let stat = await fs.statAsync(this.filepath);
        if (!stat || !stat.isFile())
            return null;
        let content = await fs.readFile(this.filepath, 'utf8');
        if (!content.trim())
            return {};
        try {
            return JSON.parse(content);
        }
        catch (e) {
            return null;
        }
    }
    async clear() {
        let stat = await fs.statAsync(this.filepath);
        if (!stat || !stat.isFile())
            return;
    }
    async destroy() {
        await fs.unlinkAsync(this.filepath);
    }
}
exports.default = DB;
